x-pack/plugins/apm/server/routes/traces/get_aggregated_critical_path.ts

```diff
@@ -0,0 +1,402 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ProcessorEvent } from '@kbn/observability-plugin/common';
+import { rangeQuery, termsQuery } from '@kbn/observability-plugin/server';
+import { Logger } from '@kbn/logging';
+import {
+  AGENT_NAME,
+  PROCESSOR_EVENT,
+  SERVICE_NAME,
+  SPAN_NAME,
+  SPAN_SUBTYPE,
+  SPAN_TYPE,
+  TRACE_ID,
+  TRANSACTION_NAME,
+  TRANSACTION_TYPE,
+} from '../../../common/elasticsearch_fieldnames';
+import { AgentName } from '../../../typings/es_schemas/ui/fields/agent';
+import { APMEventClient } from '../../lib/helpers/create_es_client/create_apm_event_client';
+
+type OperationMetadata = {
+  [SERVICE_NAME]: string;
+  [AGENT_NAME]: AgentName;
+} & (
+  | {
+      [PROCESSOR_EVENT]: ProcessorEvent.transaction;
+      [TRANSACTION_TYPE]: string;
+      [TRANSACTION_NAME]: string;
+    }
+  | {
+      [PROCESSOR_EVENT]: ProcessorEvent.span;
+      [SPAN_NAME]: string;
+      [SPAN_TYPE]: string;
+      [SPAN_SUBTYPE]: string;
+    }
+);
+
+type OperationId = string;
+
+type NodeId = string;
+
+export interface CriticalPathResponse {
+  metadata: Record<OperationId, OperationMetadata>;
+  timeByNodeId: Record<NodeId, number>;
+  nodes: Record<NodeId, NodeId[]>;
+  rootNodes: NodeId[];
+  operationIdByNodeId: Record<NodeId, OperationId>;
+}
+
+const TWO_DAYS = 2 * 24 * 60 * 60 * 1000;
+
+export async function getAggregatedCriticalPath({
+  traceIds,
+  start,
+  end,
+  apmEventClient,
+  serviceName,
+  transactionName,
+  logger,
+}: {
+  traceIds: string[];
+  start: number;
+  end: number;
+  apmEventClient: APMEventClient;
+  serviceName: string | null;
+  transactionName: string | null;
+  logger: Logger;
+}): Promise<{ criticalPath: CriticalPathResponse | null }> {
+  const now = Date.now();
+
+  const response = await apmEventClient.search('get_aggregated_critical_path', {
+    apm: {
+      events: [ProcessorEvent.span, ProcessorEvent.transaction],
+    },
+    body: {
+      size: 0,
+      track_total_hits: false,
+      query: {
+        bool: {
+          filter: [
+            ...termsQuery(TRACE_ID, ...traceIds),
+            ...rangeQuery(start - TWO_DAYS, end + TWO_DAYS),
+          ],
+        },
+      },
+      aggs: {
+        critical_path: {
+          scripted_metric: {
+            params: {
+              // can't send null parameters to ES. undefined will be removed during JSON serialisation
+              serviceName: serviceName || undefined,
+              transactionName: transactionName || undefined,
+            },
+            init_script: {
+              source: `
+                state.eventsById = [:];
+                state.metadataByOperationId = [:];
+              `,
+            },
+            map_script: {
+              source: `
+                String toHash (def item) {
+                  return item.toString();
```

Comment:
This is a very inefficient (in terms of size) hashing function, and we should improve it, but it might require additional APIs in Painless. I've opened a ticket for that here: https://github.com/elastic/kibana/issues/144225